src/lib/AuthManager.js

```diff
@@ -0,0 +1,236 @@
+  // contain the request in this AuthSession. Sets a AuthSessionId in the request.
+  fetch(url, init) {
+    init.credentials = "omit"
+    if (!this.needsAuthentication) {
+      delete init.headers.Authorization
+    }
+    init.headers.AuthSessionId = this.id
+    return fetch(url, init)
+  }
+
+  acceptCookies(header) {
+    if (!this.cookieJar.storeFromHeader(header)) {
+      if (this.everRecievedCookie) {
+        // if and only if our browser grants us access to httpOnly cookies,
+        // _and_ nextcloud has stopped sending new cookies, stop sending login data
+        this.needsAuthentication = false
+      }
+    } else {
+      this.everRecievedCookie = true
+    }
+  }
+}
+
+// Handles listeners for AuthSessions. Should have exactly one global instance.  
+export class AuthManager {
+
+  constructor() {
+    this.sessions = {} // dict: AuthSessionId => AuthSession
+    this.requests = {} // dict: request.requestId => AuthSessionId
+    this.sessionIdState = 0
+
+    this.onResponseListener = e => this.onResponse(e)
+    this.onRequestListener = e => this.onRequest(e)
+  }
+
+  updateListeners() {
+    let filter = this.getRequestFilter()
+    if (filter.urls.length <= 0) {
+      this.removeListeners()
+    } else {
+      this.addListeners(filter)
+    }
+  }
+
+  // adds or overwrites listeners
+  addListeners(filter) {
+
+    let respPerms = ["blocking", "responseHeaders"]
+    let reqPerms = ["blocking", "requestHeaders"]
+
+    // "extraHeaders" is required additionally for chrome to show relevant cookies
+    if (browser.isChrome) {
+      respPerms.push("extraHeaders")
+      reqPerms.push("extraHeaders")
+    }
+
+    browser.webRequest.onHeadersReceived.addListener(
+      this.onResponseListener,
+      filter,
+      respPerms
+    )
+
+    browser.webRequest.onBeforeSendHeaders.addListener(
+      this.onRequestListener,
+      filter,
+      reqPerms
+    )
+  }
+
+  removeListeners() {
+    browser.webRequest.onHeadersReceived.removeListener(this.onResponseListener)
+    browser.webRequest.onBeforeSendHeaders.removeListener(this.onRequestListener)
+  }
+
+  onResponse(e) {
+      let header = e.responseHeaders
+      // get authSession by request id
+      let authSessionId = this.requests[e.requestId]
+      if (authSessionId === undefined) {
+        // request wasn't handled by this
+        console.info("Detected a response not meant for the bookmarks API. Returning now.")
+      } else {
+        // save cookie in session
+        this.sessions[authSessionId].acceptCookies(header)
+        // remove request and its ownership
+        delete this.requests[e.requestId]
+        delete this.sessions[authSessionId].ownedRequests[e.requestId]
+      }
+      return {responseHeaders: e.responseHeaders};
+  }
+
+  onRequest(e) {
+      let header = e.requestHeaders
+      // get the authSessionId
+      let authSessionId
+      for (let object of header) {
+        if (object.name.toLowerCase() == "authsessionid") {
```

Comment:
this should probably be a constant and something more proprietary looking, like `x-floccus-session-id` or something similar :)